man/ibnbd.8
.TH IBNBD "8" "Februar 2020" "ibnbd 1.0.0" "System Administration Utilities"
.SH NAME
ibnbd - Configuration tool for IBNBD driver and IBTRS library.
.SH SYNOPSIS
ibnbd client|server|both|help|version ...
.SH SYNOPSIS

    --help          Help output
    --verbose       Verbose output
    --debug         Debug mode
    --simulate      Only print modifying operations, do not execute



    client          Operations of client
    server          Operations of server
    both            Operations of both client and server
    help            Display help and exit
    version         Display version of this tool and exit


ibnbd client devices|sessions|paths|help ...

    devices         Map/unmapped/modify devices
    sessions        Operate on sessions
    paths           Handle paths
    help            Display help and exit


ibnbd server devices|sessions|paths|help ...

    devices         List devices
    sessions        Operate on sessions
    paths           Handle paths
    help            Display help and exit


ibnbd both dump|help ...

    dump            Dump information about all ibnbd objects.
    help            Display help and exit


.SH DESCRIPTION
The commands of the tool are structured in the following fashion:
client vs server -> device vs. session vs. path. If the context of
a command is unambiguous, it can be also called directly.
For example: ibnbd map (instead of ibnbd client device map), ibnbd session list
(instead of ibnbd client session list), ibnbd show client@server (instead of ibnbd client session show client@server), etc.
.SH SUBCOMMANDS
.B

ibnbd client device list [OPTIONS]

List information on devices.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Session        Name of the IBTRS session of the device
                    mapping_path    Mapping Path   Mapping name of the remote device
                    devname         Device         Device name under /dev/. I.e. ibnbd0
                    devpath         Device path    Device path under /dev/. I.e. /dev/ibnbd0
                    state           State          State of the IBNBD device. (client only)
                    access_mode     Access Mode    RW mode of the device: ro, rw or migration
                    rx_sect         RX             Amount of data read from the device
                    tx_sect         TX             Amount of data written to the device
                    direction       Direction      Direction of data transfer: imported or exported

                    Default: sessname,mapping_path,devname,state,access_mode

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    notree          Don't display paths for each sessions
    noheaders       Don't print headers
    nototals        Don't print totals
    help            Display help and exit. [fields|all]


.B

ibnbd client device show <device> [OPTIONS]

Show information about an ibnbd block device.

Arguments:
    <device>        Name of a local or a remote block device.
                    I.e. ibnbd0, /dev/ibnbd0, d12aef94-4110-4321-9373-3be8494a557b.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Session        Name of the IBTRS session of the device
                    mapping_path    Mapping Path   Mapping name of the remote device
                    devname         Device         Device name under /dev/. I.e. ibnbd0
                    devpath         Device path    Device path under /dev/. I.e. /dev/ibnbd0
                    state           State          State of the IBNBD device. (client only)
                    access_mode     Access Mode    RW mode of the device: ro, rw or migration
                    rx_sect         RX             Amount of data read from the device
                    tx_sect         TX             Amount of data written to the device
                    direction       Direction      Direction of data transfer: imported or exported

                    Default: sessname,mapping_path,devname,state,access_mode

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    help            Display help and exit. [fields|all]


.B

ibnbd client device map <device> from <server> [OPTIONS]

Map a device from a given server

Arguments:
    <device>        Path to the device to be mapped on server side
    from <server>   Address, hostname or session name of the server

Options:
    <path>          Path(s) to establish: [src_addr@]dst_addr
                    Address is [ip:]<ipv4>, [ip:]<ipv6> or gid:<gid>
    {rw}            Access permission on server side: ro|rw|migration. Default: rw
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd client device resize <device> <size> [OPTIONS]

Change size of a mapped device

Arguments:
    <device>        Name of the device to be resized
    <size>          New size of the device in bytes
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E

Options:
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd client device unmap <device> [OPTIONS]

Umap a given imported device

Arguments:
    <device>        Name of the device to be unmapped

Options:
    force           Force operation
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd client device remap <device> [OPTIONS]

Unmap and map again an imported device

Arguments:
    <identifier>    Identifier of a device to be remapped.

Options:
    force           Force operation
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd client session list [OPTIONS]

List information on sessions.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Session name   Name of the session
                    path_cnt        Path cnt       Number of paths
                    act_path_cnt    Act path cnt   Number of active paths
                    state           State          State of the session.
                    path_uu         PS             Up (U) or down (_) state of every path
                    mp              MP Policy      Multipath policy
                    mp_short        MP             Multipath policy (short)
                    rx_bytes        RX             Bytes received
                    tx_bytes        TX             Bytes send
                    inflights       Inflights      Inflights
                    reconnects      Reconnects     Reconnects
                    side            Direction      Direction of the session: incoming or outgoing

                    Default: sessname,state,path_uu,mp_short,tx_bytes,rx_bytes,inflights,reconnects

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    notree          Don't display paths for each sessions
    noheaders       Don't print headers
    nototals        Don't print totals
    help            Display help and exit. [fields|all]


.B

ibnbd client session show <session> [OPTIONS]

Show information about an ibnbd session.

Arguments:
    <session>       Session name or remote hostname.
                    I.e. ps401a-1@st401b-2, st401b-2, <ip1>@<ip2>, etc.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Session name   Name of the session
                    path_cnt        Path cnt       Number of paths
                    act_path_cnt    Act path cnt   Number of active paths
                    state           State          State of the session.
                    path_uu         PS             Up (U) or down (_) state of every path
                    mp              MP Policy      Multipath policy
                    mp_short        MP             Multipath policy (short)
                    rx_bytes        RX             Bytes received
                    tx_bytes        TX             Bytes send
                    inflights       Inflights      Inflights
                    reconnects      Reconnects     Reconnects
                    side            Direction      Direction of the session: incoming or outgoing

                    Default: sessname,state,path_uu,mp_short,tx_bytes,rx_bytes,inflights,reconnects

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    help            Display help and exit. [fields|all]


.B

ibnbd client session remap <session> [OPTIONS]

Unmap and map again all devices of a given session

Arguments:
    <session>       Identifier of a session to remap all devices on.

Options:
    force           Force operation
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd client session reconnect <session> [OPTIONS]

Disconnect and connect again a whole session

Arguments:
    <session>       Name or identifier of a session.

Options:
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd client path list [OPTIONS]

List information on paths.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Sessname       Name of the session.
                    pathname        Path name      Path name
                    src_addr        Client Addr    Client address of the path
                    src_addr_s      Client Addr    Normalized client address.
                    dst_addr        Server Addr    Server address of the path
                    dst_addr_s      Server Addr    Normalized server address.
                    hca_name        HCA            HCA name
                    hca_port        Port           HCA port
                    state           State          Name of the path
                    rx_bytes        RX             Bytes received
                    tx_bytes        TX             Bytes send
                    inflights       Inflights      Inflights
                    reconnects      Reconnects     Reconnects
                    direction       Direction      Direction of the path: incoming or outgoing

                    Default: sessname,hca_name,hca_port,dst_addr_s,state,tx_bytes,rx_bytes,inflights,reconnects

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    notree          Don't display paths for each sessions
    noheaders       Don't print headers
    nototals        Don't print totals
    help            Display help and exit. [fields|all]


.B

ibnbd client path show <path> [OPTIONS]

Show information about an ibnbd transport path.

Arguments:
    <path>          In order to display path information, path name or identifier
                    has to be provided, i.e. st401b-2:1.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Sessname       Name of the session.
                    pathname        Path name      Path name
                    src_addr        Client Addr    Client address of the path
                    src_addr_s      Client Addr    Normalized client address.
                    dst_addr        Server Addr    Server address of the path
                    dst_addr_s      Server Addr    Normalized server address.
                    hca_name        HCA            HCA name
                    hca_port        Port           HCA port
                    state           State          Name of the path
                    rx_bytes        RX             Bytes received
                    tx_bytes        TX             Bytes send
                    inflights       Inflights      Inflights
                    reconnects      Reconnects     Reconnects
                    direction       Direction      Direction of the path: incoming or outgoing

                    Default: sessname,hca_name,hca_port,dst_addr_s,state,tx_bytes,rx_bytes,inflights,reconnects

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    help            Display help and exit. [fields|all]


.B

ibnbd client path disconnect <session> [OPTIONS]

Disconnect a path a given session

Arguments:
    <identifier>    Name or identifier of of a path:
                    [pathname], [sessname:port], etc.

Options:
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd client path reconnect <path> [OPTIONS]

Disconnect and connect again a single path of a session

Arguments:
    <identifier>    Name or identifier of a path:
                    [pathname], [sessname:port], etc.

Options:
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd session add <session> <path> [OPTIONS]

Add a new path to an existing session

Arguments:
    <session>       Name of the session to add the new path to
    <path>          Path to be added: [src_addr@]dst_addr
                    Address is of the form ip:<ipv4>, ip:<ipv6> or gid:<gid>

Options:
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd client path delete <path> [OPTIONS]

Delete a given path from the corresponding session

Arguments:
    <path>          Name or any unique identifier of a path:
                    [pathname], [sessname:port], etc.

Options:
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd client path readd <path> [OPTIONS]

Delete and add again a given path to the corresponding session

Arguments:
    <path>          Name or any unique identifier of a path:
                    [pathname], [sessname:port], etc.

Options:
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd server device list [OPTIONS]

List information on devices.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Session        Name of the IBTRS session of the device
                    mapping_path    Mapping Path   Mapping name of the remote device
                    devname         Device         Device name under /dev/. I.e. ibnbd0
                    devpath         Device path    Device path under /dev/. I.e. /dev/ibnbd0
                    state           State          State of the IBNBD device. (client only)
                    access_mode     Access Mode    RW mode of the device: ro, rw or migration
                    rx_sect         RX             Amount of data read from the device
                    tx_sect         TX             Amount of data written to the device
                    direction       Direction      Direction of data transfer: imported or exported
                    Default: sessname,mapping_path,devname,access_mode

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    notree          Don't display paths for each sessions
    noheaders       Don't print headers
    nototals        Don't print totals
    help            Display help and exit. [fields|all]


.B

ibnbd server device show <device> [OPTIONS]

Show information about an ibnbd block device.

Arguments:
    <device>        Name of a local or a remote block device.
                    I.e. ibnbd0, /dev/ibnbd0, d12aef94-4110-4321-9373-3be8494a557b.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Session        Name of the IBTRS session of the device
                    mapping_path    Mapping Path   Mapping name of the remote device
                    devname         Device         Device name under /dev/. I.e. ibnbd0
                    devpath         Device path    Device path under /dev/. I.e. /dev/ibnbd0
                    state           State          State of the IBNBD device. (client only)
                    access_mode     Access Mode    RW mode of the device: ro, rw or migration
                    rx_sect         RX             Amount of data read from the device
                    tx_sect         TX             Amount of data written to the device
                    direction       Direction      Direction of data transfer: imported or exported
                    Default: sessname,mapping_path,devname,access_mode

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    help            Display help and exit. [fields|all]


.B

ibnbd server session list [OPTIONS]

List information on sessions.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Session name   Name of the session
                    path_cnt        Path cnt       Number of paths
                    act_path_cnt    Act path cnt   Number of active paths
                    state           State          State of the session.
                    path_uu         PS             Up (U) or down (_) state of every path
                    mp              MP Policy      Multipath policy
                    mp_short        MP             Multipath policy (short)
                    rx_bytes        RX             Bytes received
                    tx_bytes        TX             Bytes send
                    inflights       Inflights      Inflights
                    reconnects      Reconnects     Reconnects
                    side            Direction      Direction of the session: incoming or outgoing
                    Default: sessname,path_cnt,tx_bytes,rx_bytes,inflights

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    notree          Don't display paths for each sessions
    noheaders       Don't print headers
    nototals        Don't print totals
    help            Display help and exit. [fields|all]


.B

ibnbd server session show <session> [OPTIONS]

Show information about an ibnbd session.

Arguments:
    <session>       Session name or remote hostname.
                    I.e. ps401a-1@st401b-2, st401b-2, <ip1>@<ip2>, etc.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Session name   Name of the session
                    path_cnt        Path cnt       Number of paths
                    act_path_cnt    Act path cnt   Number of active paths
                    state           State          State of the session.
                    path_uu         PS             Up (U) or down (_) state of every path
                    mp              MP Policy      Multipath policy
                    mp_short        MP             Multipath policy (short)
                    rx_bytes        RX             Bytes received
                    tx_bytes        TX             Bytes send
                    inflights       Inflights      Inflights
                    reconnects      Reconnects     Reconnects
                    side            Direction      Direction of the session: incoming or outgoing
                    Default: sessname,path_cnt,tx_bytes,rx_bytes,inflights

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    help            Display help and exit. [fields|all]


.B

ibnbd server session disconnect <session> [OPTIONS]

Disconnect all paths on a given session

Arguments:
    <session>       Name or identifier of a session.

Options:
    verbose         Verbose output
    help            Display help and exit


.B

ibnbd server path list [OPTIONS]

List information on paths.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Sessname       Name of the session.
                    pathname        Path name      Path name
                    src_addr        Client Addr    Client address of the path
                    src_addr_s      Client Addr    Normalized client address.
                    dst_addr        Server Addr    Server address of the path
                    dst_addr_s      Server Addr    Normalized server address.
                    hca_name        HCA            HCA name
                    hca_port        Port           HCA port
                    state           State          Name of the path
                    rx_bytes        RX             Bytes received
                    tx_bytes        TX             Bytes send
                    inflights       Inflights      Inflights
                    reconnects      Reconnects     Reconnects
                    direction       Direction      Direction of the path: incoming or outgoing
                    Default: sessname,hca_name,hca_port,src_addr_s,tx_bytes,rx_bytes,inflights

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    notree          Don't display paths for each sessions
    noheaders       Don't print headers
    nototals        Don't print totals
    help            Display help and exit. [fields|all]


.B

ibnbd server path show <path> [OPTIONS]

Show information about an ibnbd transport path.

Arguments:
    <path>          In order to display path information, path name or identifier
                    has to be provided, i.e. st401b-2:1.

Options:
    {fields}        Comma separated list of fields to be printed. The list can be
                    prefixed with '+' or '-' to add or remove fields from the
                    default selection.

                    Field           Header         Description
                    sessname        Sessname       Name of the session.
                    pathname        Path name      Path name
                    src_addr        Client Addr    Client address of the path
                    src_addr_s      Client Addr    Normalized client address.
                    dst_addr        Server Addr    Server address of the path
                    dst_addr_s      Server Addr    Normalized server address.
                    hca_name        HCA            HCA name
                    hca_port        Port           HCA port
                    state           State          Name of the path
                    rx_bytes        RX             Bytes received
                    tx_bytes        TX             Bytes send
                    inflights       Inflights      Inflights
                    reconnects      Reconnects     Reconnects
                    direction       Direction      Direction of the path: incoming or outgoing
                    Default: sessname,hca_name,hca_port,src_addr_s,tx_bytes,rx_bytes,inflights

    {format}        Output format: csv|json|xml
    {unit}          Units to use for size (in binary): B|K|M|G|T|P|E
    help            Display help and exit. [fields|all]


.B

ibnbd server path disconnect <session> [OPTIONS]

Disconnect a path a given session

Arguments:
    <identifier>    Name or identifier of of a path:
                    [pathname], [sessname:port], etc.

Options:
    verbose         Verbose output
    help            Display help and exit


.SH EXAMPLES
List server devices

.B ibnbd server devices list

List client sessions

.B ibnbd client sessions list 

List paths of server, display sizes in KB, display all columns


.B ibnbd server paths list K all

List imported on client devices, show only mapping_path and devpath, output in json

.B ibnbd client devices list mapping_path,devpath json

.SH COPYRIGHT
Copyright \(co 2019 - 2020 IONOS Cloud GmbH. All Rights Reserved
.SH AUTHORS
Danil Kipnis <danil.kipnis@cloud.ionos.com>
.RE
Lutz Pogrell <lutz.pogrell@cloud.ionos.com>


ibnbd.h2m.sh
#!/bin/bash
# SPDX-License-Identifier: GPL-2.0-or-later

########################################
# Run that and redirect to man/ibnbd.8 #
########################################

PATH=$PATH:./
VER=$(egrep -oe "[0-9,.]+" -m 1 debian/changelog)
DATE=$(date +"%B %Y")

echo ".TH IBNBD \"8\" \"$DATE\" \"ibnbd $VER\" \"System Administration Utilities\""
echo ".SH NAME"
echo "ibnbd - Configuration tool for IBNBD driver and IBTRS library."
echo ".SH SYNOPSIS"

ibnbd help | grep 'Usage:' | sed 's/Usage: //g'

echo ".SH SYNOPSIS"
ibnbd help all | grep -v "client|server|both|help" | grep -v -e "--help|--verbose|--debug|--simulate" | sed 's/Usage: //g'

echo -e "\n"
echo ".SH DESCRIPTION"
echo "The commands of the tool are structured in the following fashion:"
echo "client vs server -> device vs. session vs. path. If the context of"
echo "a command is unambiguous, it can be also called directly."
echo "For example: ibnbd map (instead of ibnbd client device map), ibnbd session list"
echo "(instead of ibnbd client session list), ibnbd show client@server (instead of ibnbd client session show client@server), etc."

modes="client server both"
objects="device session path"
cmds="list show map resize unmap remap disconnect reconnect add delete readd"

echo ".SH SUBCOMMANDS"

for m in $modes; do
	for o in $objects; do
		for c in $cmds; do
			if ibnbd $m $o $c help > /dev/null; then
				echo -e ".B\n"
				ibnbd $m $o $c help all| sed 's/[[:space:]]*$//' | sed 's/Usage: //'
				echo -e "\n"
			fi
		done
	done
done

echo ".SH EXAMPLES"
echo -e "List server devices\n"
echo -e ".B ibnbd server devices list\n"
echo -e "List client sessions\n"
echo -e ".B ibnbd client sessions list \n"
echo -e "List paths of server, display sizes in KB, display all columns\n\n"
echo -e ".B ibnbd server paths list K all\n"
echo -e "List imported on client devices, show only mapping_path and devpath, output in json\n"
echo -e ".B ibnbd client devices list mapping_path,devpath json\n"

echo ".SH COPYRIGHT"
echo "Copyright \(co 2019 - 2020 IONOS Cloud GmbH. All Rights Reserved"
echo ".SH AUTHORS"
echo "Danil Kipnis <danil.kipnis@cloud.ionos.com>"
echo ".RE"
echo "Lutz Pogrell <lutz.pogrell@cloud.ionos.com>"
